### ISSUE TITLE
Implement batch drawing on WebGL

### ISSUE THREAD

**Issue by agmcleod** [melonjs/melonJS#591]
This should reduce heavy usage of buffers, and number of draw calls required. Leading to a performance boost.

**Comment by parasyte** [comment]
About the save/restore performance, it will be better to move these operations into the methods on the Renderer class. I'll provide an example and describe my thoughts on the issue and how this change will make it better.

If we look at the particle class, we can see right away that each particle must save and restore the context state around every `drawImage` call: https://github.com/melonjs/melonJS/blob/640e5038ccad75f0c4edbca5c8408f2fc0f99bab/src/particles/particle.js#L133-L155 This is just one example, but it highlights a worst-case scenario where hundreds of images are being drawn, each saving and restoring the context. This is actually necessary for Canvas 2D operations to work correctly, since it only has a shared context. However, WebGL does not have this limitation, and we shouldn't emulate it!

Instead, we can push any context information we want to the GPU with WebGL on a per-triangle basis. It is therefore more efficient to send the required contextual information as part of the `drawImage` call, rather than maintaining a stack and globally shared context.

The `drawImage` method signature will change like this:

`drawImage(Image|Object, sx, sy, sw, sh, dx, dy, dw, dh)`

Where `Object` is a key-value pair that provides additional contextual information like blend color, transformations, and whether to update or preserve the global context:

```javascript
{
  "image" : image,
  "transform" : transform,
  "update" : false
}
```

`update=false` is default, which will cause the CanvasRenderer to wrap the `drawImage` call in `save/restore`. And in WebGLRenderer, nothing special will happen. `update=true` is the opposite; nothing happens in `CanvasRenderer`, and the global context gets updated by WebGLRenderer.

Here's a list of settings available:

**Required**:

* `texture` : a Texture instance OR...
  * `image` : an Image instance

**Optional**:

* `transform` : a Matrix2d instance OR...
  * `pos` : a Vector2d instance AND...
  * `angle` : a Number [0..2pi] AND...
  * `scale` : a Vector2d instance
* `color` : a Color instance OR...
  * `alpha` : a Number [0..1]

We will need the same kind of API update for the `fill*` and `stroke*` methods.

This proposed change will reduce unnecessary overhead in WebGLRenderer by removing the global context emulation. Updating the global context is opt-in, and will no longer require a stack for general purpose image rendering. With these changes, we can get rid of some silly workarounds like setting the global context color to white at the end of drawing: https://github.com/melonjs/melonJS/commit/080c87d8219d6e2521ebb7515534295881389bcb

**Comment by obiot** [comment]
it's hard to provide negative comments, since you have been on your own on this one for a couple of weeks now, but to be honest I'm not sure I really like your last idea as in my opinion, you should provide the same API either for canvas and/or webgl, and that should keep being transparent for the final end user.

However I'm not sure why you propose this change, as today the `save/restore` functions are called throuhg the `renderers` and that should then be managed from there (with the webgl one being basically an empty function),

**Comment by parasyte** [comment]
@obiot Sorry for not being clear. The API will be the same on both renderers. The internal workings will be quite different, though. What I really want to do here is remove the need to call save/restore at all, even by code like the particle emitter (linked above) and the animation sheet class.

Preserving the semantic of a global context state is a bad idea, in my honest opinion. WebGL doesn't need it, and we can hide it entirely for the CanvasRenderer. In other words, the renderer API should reflect the best case scenario; it should expose everything we can do with WebGL, and not abstract it away to look like Canvas 2D.

**Comment by obiot** [comment]
oh I see, so then yes I think it makes sense :P

**Comment by obiot** [comment]
FYI, under FF 28, the shader compilation fails with the following message :
````javascript
me.video.Error: ERROR: 0:10: break disallowed outside switch/loop body
ERROR: 0:17: break disallowed outside switch/loop body
ERROR: 0:24: break disallowed outside switch/loop body
ERROR: 0:31: break disallowed outside switch/loop body
ERROR: 0:38: break disallowed outside switch/loop body
ERROR: 0:45: break disallowed outside switch/loop body
ERROR: 0:52: break disallowed outside switch/loop body
ERROR: 0:59: break disallowed outside switch/loop body
ERROR: 0:66: break disallowed outside switch/loop body
ERROR: 0:73: break disallowed outside switch/loop body
ERROR: 0:80: break disallowed outside switch/loop body
ERROR: 0:87: break disallowed outside switch/loop body
ERROR: 0:94: break disallowed outside switch/loop body
ERROR: 0:101: break disallowed outside switch/loop body
ERROR: 0:108: break disallowed outside switch/loop body
ERROR: 0:115: break disallowed outside switch/loop body
````

**Comment by ldd** [comment]
Under Firefox 34, on windows 8.1 everything works (after removing the comments in the shaders) with the comments there, I get a similar error.

**Comment by parasyte** [comment]
Yeah, there are problems in this shader. The issue with the comment is in the grunt-replace task; it doesn't remove comments properly on Windows: https://github.com/melonjs/melonJS/blob/389e6435e75c67c2099e9441b898f27f3fafcdc5/Gruntfile.js#L87-L88 I suspect that added `$` to the double-slash side of the RegExp will fix that, but I am unable to test on Windows.

The problem with FireFox 28 (really old browser, BTW!) looks like it is optimizing the GLSL by unrolling the loop (GOOD!) but the break statement confuses it in the unrolled output.

I get similar weird behavior with this shader when passing it to [glsl-optimizer](https://github.com/parasyte/grunt-glsl-optimizer). It fails to compile unless I replace the "dynamic index" with a constant. This is because glsl-optimizer does not unroll the loop (BAD!)

The obvious thing would be to run it through an optimizer that unrolls the loop and strips comments and whitespace, etc. glsl-optimizer isn't doing that for us, so I might have to just write a dumb parser specifically for this.

Also take note of the comment in the fragment shader; to get the most out of the GPU, the loop needs to be sized appropriately for the GPU. I have it hardcoded to use the number of texture units in my laptop's GPU at the moment. Unrolling the loop would have to be done at runtime to make that happen.

**Comment by obiot** [comment]
Indeed FF28 is old, but i had the same issue after updating to 31, on my side it was on OSX though,m (but does not certainely change anything)

**Comment by parasyte** [comment]
@obiot The fragment shader will be replaced. But FTR what GPU is in your Mac?

**Comment by obiot** [comment]
MBA Intel HD5000 :)

**Comment by obiot** [comment]
i just tried on my windows machine (FF34, Intel HD4000) and firefox just crahsed when I tried to open the platformer :P:P:P

**Comment by parasyte** [comment]
Lots of stuff happening here. :smiley: Finally I replaced the multiple `bufferSubData` calls with a single call. Next I will split the stream buffer into two buffers; one that is updated often (vertices and texture coordinates) and one that it update infrequently (color and texture index). Technically the color will change often when objects are fading and such. So we'll have to do some tuning on this stuff. But this should be a good start.

**Comment by parasyte** [comment]
I added the separate buffers patch to a new branch: https://github.com/melonjs/melonJS/compare/experimental/WebGL_static_buffer (See previous commit log for details)

This new code requires some additional CPU time for hashing the static buffer. Hashing should be the fastest way to check whether the static attributes for a quad have changed... Although to be honest I haven't tried a naïve approach with a ton of if-statements for each array element. ;) Something tells me that won't perform quite as well.

The additional CPU time required is:

* 0.25% with 140 quads.
* 0.56% with 256 quads.
* 1.02% with 496 quads.

The scale is roughly linear, so expect 32% additional CPU usage for 16,000 quads (the current limit for a single batch operation). It's unclear if this additional CPU overhead is worth the GPU bandwidth reduction. I'll maintain this new branch in parallel with master. At least until it can be determined if it's useful or not.

**Comment by parasyte** [comment]
@ldd The fragment shader should now build properly on your environment. The trick was using the `preserveOrder` option in the grunt-replace task. :smiley:

**Comment by obiot** [comment]
yep good job ! I confirm it fixes the build issue at least on my MBA (using Firefox), will try tomorrow on my windows machine.

unrelated, but since I did a `nom install` the jasmine task is now however failing :
````
Running "jasmine:src" (jasmine) task
Testing jasmine specs via PhantomJS

>> Error caught from PhantomJS. More info can be found by opening the Spec Runner in a browser.
Warning: SyntaxError: Parse error Use --force to continue.

Aborted due to warnings.
````
and I don't see anything in the spec runner. Do you also have that issue ?

**Comment by parasyte** [comment]
Yeah, I do see that now. It didn't happen in my earlier tests last night. But I got the build failure notice from Travis-CI this morning.

Working on it!

**Comment by parasyte** [comment]
Alright, I think it's finally in a pretty stable state! The stuff I did tonight focuses on customizability of the WebGL environment. I don't want to tie any users down in regards to how they use WebGL. So now it's possible to use an entirely custom Compositor class by passing the `compositor` option to `me.video.init()`! I don't think I'll be writing a different compositor any time soon, but I like that we can provide this flexibility for others who wish to experiment.

Another important change is that the `me.video.shader.createShader()` method is now independent of the singleton, allowing it to compile multiple shader programs. This I will be using in our default compositor for the line-rendering (e.g. `fillStroke`) shader program. The compositor just needs to flush when switching between shader programs.

Most of my TODO lists are already done, which is exciting! And I heard today from @ldd that his tests show a nice improvement in rendering speed. Apparently in his tests, CanvasRenderer is capable of 142 objects max and WebGLRenderer is capable of over 500. This is a good start, but I want more! :)

**Comment by parasyte** [comment]
With the last few commits, stroke (line rendering) is finally in place. It's not efficient, though. At first glance, it appears that the depth buffer can make it very efficient; we just need a way to get the Z-coordinate information into the compositor. That will likely depend on the work in #637

In the meantime, there are a few FIXME comments that need to be addressed (especially with how the uniform variables are set, and the attribute bindings are handled).

Second to that, getting fonts working (and in particular replacing the RTT thing in the debugPanel) is a priority for release. There's also a weird ghosting effect seen on the debugPanel with WebGLRenderer. That needs to be investigated further.

**Comment by parasyte** [comment]
Started working on font support in WebGL. The hack in the branch is pretty ugly, but it does make the me.Font API consistent! (Solves #619)

It's currently *very* slow with the font_text example, because it spends most of its time creating and uploading massive textures. :laughing: The secondary texture cache (proposed in the commit) will help that a little bit.

A better way to support fonts in WebGL will be important long-term, but this will work for 2.1!

**Comment by agmcleod** [comment]
Awesome! For now we can recommend keeping usage of the me.Font api simple, or to use Canvas instead :)

**Comment by obiot** [comment]
@parasyte if you don't mind, could you maybe create one or several small tickets to better identify what's left to be done for this one ?

**Comment by parasyte** [comment]
Everything left to do is a task [here](https://github.com/melonjs/melonJS/issues/591#issuecomment-68298086) and [here](https://github.com/melonjs/melonJS/issues/591#issuecomment-68342849)

**Comment by obiot** [comment]
oh sorry, missed that, but for my defense this ticket is super long now ;P

**Comment by obiot** [comment]
did you guys see that ? 
http://patriciogonzalezvivo.com/2015/thebookofshaders/

**Comment by parasyte** [comment]
Closing this. Followup ticket is #637
